src/main/java/net/neoforged/neoforge/network/negotiation/NetworkComponentNegotiator.java

```diff
@@ -0,0 +1,182 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import net.minecraft.network.chat.Component;
+import net.minecraft.resources.ResourceLocation;
+import org.apache.commons.compress.utils.Lists;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.VisibleForTesting;
+
+/**
+ * Negotiates the network components between the server and client.
+ */
+@ApiStatus.Internal
+public class NetworkComponentNegotiator {
+
+    /**
+     * Negotiates the network components between the server and client.
+     * <p>
+     * The following rules are followed:
+     * <ul>
+     * <li>Any component that is optional on the client but is not present on the server is removed from the client's list.</li>
+     * <li>Any component that is optional on the server but is not present on the client is removed from the server's list.</li>
+     * <li>If the client has none optional components that are not present on the server, then negotiation fails</li>
+     * <li>If the server has none optional components that are not present on the client, then negotiation fails</li>
+     * <li>For each of the matching channels the following is executed:</li>
+     * <ul>
+     * <li>Check if packet flow directions are set, and if at least one is set match it to the other, by missing or wrong value fail the negotiation.</li>
+     * <li>Check if both sides have the same version, or none set.</li>
+     * </ul>
+     * <li>At this point the channels are considered compatible, pick the servers version. It does not matter what side is picked since either both have the same version, or no version at all.</li>
+     * </ul>
+     * </p>
+     * <p>
+     * If negotiation succeeds then a list of agreed upon channels and their versions is returned.
+     * </p>
+     * <p>
+     * If negotiation fails then a {@link Component} is returned with the reason for failure.
+     * </p>
+     *
+     * @param server The list of server components that the server wishes to use for communication.
+     * @param client The list of client components that the client wishes to use for communication.
+     * @return A {@link NegotiationResult} that contains the agreed upon channels and their versions if negotiation succeeded, or a {@link Component} with the reason for failure if negotiation failed.
+     */
+    public static NegotiationResult negotiate(List<NegotiableNetworkComponent> server, List<NegotiableNetworkComponent> client) {
+        //Ensure the inputs are modifiable
+        server = new ArrayList<>(server);
+        client = new ArrayList<>(client);
+
+        List<NegotiableNetworkComponent> finalServer = server;
+        final List<NegotiableNetworkComponent> disabledOptionalOnClient = client.stream()
+                .filter(NegotiableNetworkComponent::optional)
+                .filter(c -> finalServer.stream().noneMatch(c2 -> c2.id().equals(c.id())))
+                .toList();
```

Comment:
Could this be moved into a private helper method to calculate the `disabledOptional`? That way there isn't the duplication between calculating the ones for the client and the ones for the server, and on top of that because the lists would be passed as parameters to the method they wouldn't be getting modified and then we don't need the `finalServer` and `finalClient` variables which should hopefully improve the overall readability of this method.